File: 807-max-increase-to-keep-city-skyline/807-max-increase-to-keep-city-skyline.java
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
    private ArrayList<Integer> maxColumns = new ArrayList<>();

    public int maxIncreaseKeepingSkyline(int[][] grid) {
        int sum = 0;
        for (int i = 0; i < grid.length; i++) {
            int maxRow = grid[i][0];
            for (int x : grid[i]) {
                if (maxRow < x)
                    maxRow = x;
            }
            for (int j = 0; j < grid[0].length; j++) {
                if (i == 0) {
                    int max = -1;
                    for (int[] ints : grid) {
                        if (max < ints[j]) {
                            max = ints[j];
                        }
                    }
                    maxColumns.add(max);
                }
                sum += (Math.min(maxRow, maxColumns.get(j)) - grid[i][j]);
            }
        }
        return sum;
    }
}

File: 807-max-increase-to-keep-city-skyline/807-max-increase-to-keep-city-skyline.kt
import kotlin.math.min

class Solution {
    private val maxColumns = mutableListOf<Int>()
    fun maxIncreaseKeepingSkyline(grid: Array<IntArray>): Int {
        var sum = 0
        for (i in 0..grid.lastIndex) {
            val maxRow = grid[i].max() ?: 0
            for (j in 0..grid[0].lastIndex) {
                if (i == 0) {
                    var maxColumn = -1
                    for (x in 0..grid.lastIndex) {
                        if (grid[x][j] > maxColumn)
                            maxColumn = grid[x][j]
                    }
                    maxColumns.add(maxColumn)
                }
                sum += (min(maxRow, maxColumns[j]) - grid[i][j])
            }
        }
        return sum
    }
}